jaxtyping/_import_hook.py

```diff
@@ -88,14 +92,71 @@ def _dot_lookup(*elements):
     return out
 
 
-def _str_lookup(string):
-    module = ast.parse(string)
-    (expr,) = module.body
-    return expr.value
+class _Typechecker:
+    lookup = {}
+
+    def __init__(self, typechecker):
+        self.typechecker = typechecker
+
+        hash_string = lambda x: str(hashlib.md5(x.encode("utf-8")).hexdigest())
+
+        if isinstance(typechecker, str):
+            # If the typechecker is a string, we make a decorator ourselves
+            # and pass it to lookup
+            string_to_eval = (
+                "def f(x, *args, **kwargs):\n"
+                + f"  import {typechecker.split('.', 1)[0]}\n"
+                + f"  return {typechecker}(x, *args, **kwargs)"
+            )
+            local_vars = {}
+            exec(string_to_eval, {}, local_vars)
+            self.decorator = local_vars["f"]
+            self.hash = hash_string(typechecker)
+
+        elif callable(typechecker):
+            # If the typechecker is a function, treat it as a decorator
+            self.decorator = typechecker
+            try:
+                self.hash = hash_string(
+                    str(typechecker.__module__) + getsource(typechecker)
+                )
+            except OSError:
+                # OSError means the function was defined in REPL
+                # We want to miss the importlib cache, so let's use the default hash
+                self.hash = str(typechecker.__hash__())
+        elif typechecker is None:
+            # If it is None, ignore it silently (use dummy decorator)
+            self.decorator = lambda x, *_, **__: x
```

Comment:
Likewise I don't think `self.decorator` is used anywhere, and doesn't need to be saved as an attribute.

Reply:
Indeed, thank you!